marathon-bigip-ctlr.py

```diff
@@ -263,6 +263,81 @@ def set_iapp_option(x, k, v):
 logger = logging.getLogger('controller')
 
 
+def healthcheck_timeout_calculate(data):
+    """Calculate a BIG-IP Health Monitor timeout.
+
+    Args:
+        data: BIG-IP config dict
+    """
+    # Calculate timeout
+    # See the f5 monitor docs for explanation of settings:
+    # https://goo.gl/JJWUIg
+    # Formula to match up the cloud settings with f5 settings:
+    # (( maxConsecutiveFailures - 1) * intervalSeconds )
+    # + timeoutSeconds + 1
+    timeout = (
+        ((data['maxConsecutiveFailures'] - 1) * data['intervalSeconds']) +
+        data['timeoutSeconds'] + 1
+    )
+    return timeout
+
+
+def get_protocol(protocol):
+    """Return the protocol (tcp or udp)."""
+    if str(protocol).lower() == 'tcp':
+        return 'tcp'
+    if str(protocol).lower() == 'http':
+        return 'tcp'
+    if str(protocol).lower() == 'udp':
+        return 'udp'
+    return None
+
+
+def is_label_data_valid(app):
+    """Validate the Marathon app's label data.
+
+    Args:
+        app: The app to be validated
+    """
+    is_valid = True
+    msg = 'Application label {0} for {1} contains an invalid value: {2}'
+
+    # Validate mode
+    if get_protocol(app.mode) is None:
+        logger.error(msg.format('F5_MODE', app.appId, app.mode))
+        is_valid = False
+
+    # Validate port
+    if app.servicePort < 1 or app.servicePort > 65535:
+        logger.error(msg.format('F5_PORT', app.appId, app.servicePort))
+        is_valid = False
+
+    # Validate address
+    if app.bindAddr is not None:
+        try:
+            ipaddress.ip_address(app.bindAddr)
+        except ValueError:
+            logger.error(msg.format('F5_BIND_ADDR', app.appId, app.bindAddr))
+            is_valid = False
+
```

Comment:
Should we check app.balance here, too?  So that we can attribute the error to the particular label?